### ISSUE TITLE
Please start writing the glyph note as CDATA, or accept an lxml.CDATA instance instead of just str

### ISSUE THREAD

**Issue by ctrlcctrlv** [fonttools/fonttools#2745]
https://github.com/fonttools/fonttools/blob/171670a15ffca63acb8346c1c76ebb023ea3f08c/Lib/fontTools/ufoLib/glifLib.py#L855

This causes FontForge to fail to roundtrip because it's not standards-compliant and reads the newlines.

Originally:
<pre lang="sfd">
Comment: "this is the apostrophe copied five times, each time rotated 72deg withershins, then everything scaled to 50%+AAoA&#xB;+99QA"
</pre>

After processing through MFEK/sfdutf7.py@c9b4d73cdd9ee0de1bc7791bcf8975dfaa355ff7 and an unreleased `sfd2ufo` (MFEK/sfdLib.py):

```xml
<?xml version='1.0' encoding='UTF-8'?>
<glyph name="asterisk" format="2">
  <advance width="396"/>
  <unicode hex="002A"/>
  <note>
this is the apostrophe copied five times, each time rotated 72deg withershins, then everything scaled to 50%


</note>
  <outline>
    <component base="quotesingle" xScale="0.5" yScale="0.5" xOffset="136.234" yOffset="368.662"/>
    <component base="quotesingle" xScale="0.154508" xyScale="0.475528" yxScale="-0.475528" yScale="0.154508" xOffset="362.128" yOffset="451.071"/>
    <component base="quotesingle" xScale="-0.404508" xyScale="0.293893" yxScale="-0.293893" yScale="-0.404508" xOffset="353.557" yOffset="691.374"/>
    <component base="quotesingle" xScale="-0.404508" xyScale="-0.293893" yxScale="0.293893" yScale="-0.404508" xOffset="122.367" yOffset="757.481"/>
    <component base="quotesingle" xScale="0.154508" xyScale="-0.475528" yxScale="0.475528" yScale="0.154508" xOffset="-11.9465" yOffset="558.034"/>
  </outline>
  <lib>
    <dict>
      <key>public.markColor</key>
      <string>0,0,0.4,1</string>
    </dict>
  </lib>
</glyph>
```

And then written back with FontForge:
```sfd
Comment: "+AAoA-this is the apostrophe copied five times, each time rotated 72deg withershins, then everything scaled to 50%+AAoACvfUAAoA"
```

**Comment by anthrotype** [comment]
IIUC, you would like to avoid glifLib inserting the leading/trailing newline characters

I'm actually not sure why these were added in the first place, probably just to make them look nicer in the XML dump?

maybe @typesupply (who authored those lines 4 years ago) remembers:

https://github.com/fonttools/fonttools/commit/5f39650425017329f0d0d81606bad59237398238

If you look at the note parsing code, it strips all leading/trailing whitespace and normalizes the newline characters as well:

https://github.com/fonttools/fonttools/blob/171670a15ffca63acb8346c1c76ebb023ea3f08c/Lib/fontTools/ufoLib/glifLib.py#L1242-L1245

The UFO spec is vague and doesn't impose any particular requirements on the `note` elements. https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#note

> This element is a place for the user to define arbitrary text about the glyph. There is no standardized structure for the note apart from that it being the content of the note element.

Maybe these things should be clarified in the spec so that implementations can all agree

**Comment by anthrotype** [comment]
yes, perhaps using CDATA would be the best way to ensure the text is retained exactly as it it was. Though I wonder if that could create other issues for other implementations that now expect note to be a plain text string. The spec right now doesn't preclude the use of CDATA ("There is no standardized structure for the note apart from that it being the content of the note element.").

**Comment by anthrotype** [comment]
one thing that's certain is that note is supposed to be _text_, not arbitrary binary data (there's the UFO `data` directory for that). And that it's supposed to be short text, not an essay... Maybe it's sufficient that the spec aligns with ufoLib's implementation and clarifies that upon serializing, a newline is inserted at the beginning and at the end of the text content (so that the `<note>` element tags are on a separate line); then upon deserializing, newlines are normalized to unix line endings `\n` and all leading/trailing whitespace is stripped from each line, empty lines are stripped too. I think this capture the spirit of the glifLib current implementation.

**Comment by ctrlcctrlv** [comment]
Please don't fix it that way. FontForge is very hard to make any change at all to, this is why I have turned my attention to sfdLib/sfdutf7/sfdnormalize.

**Comment by ctrlcctrlv** [comment]
Or, perhaps, you can do that if you want, so long as it's OK if I make the field able to receive an `lxml.CDATA` object. :-)

**Comment by anthrotype** [comment]
i;m not entirely sure what specific fix you're suggesting, you're free to propose one with a PR and can discuss in-line

**Comment by ctrlcctrlv** [comment]
yes okay, that's what i opened this for, to know if there was a big reason for it being  how it is before PR

**Comment by anthrotype** [comment]
can't you just deal with fontforge's idiosyncrasies inside sfdLib/sft2ufo and strip the unwanted leading/trailing newlines in there?

**Comment by justvanrossum** [comment]
Alternatively: don't use `note` but use a custom lib key.

**Comment by ctrlcctrlv** [comment]
Well that would still cause a roundtrip issue unless FontForge were patched which I'm trying to avoid doing going forwards.

**Comment by anthrotype** [comment]
I filed https://github.com/unified-font-object/ufo-spec/issues/210 to clarify formatting of note text.
